HPCC-22124 File services permits and defaults to copying a super index into a invalid single index with multiple TLK's

AttilaVamos opened the pull request:
Add Superfile check for a simple logical file copy.

Add supercopy.ecl to Regression Suite

Update 3 after review,

Fix problems caused by the merge local and remote source file handling

Remove 'foreignfdesc'

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually with ECLWatch remote copy.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-12606, label: HPCC-22124-fix-master starts now.
The reason of this test is: Did not build yet.
Commit ID: 233c37c0b8bcfe99ebba2e8b762cf2ab72f32751
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 233c37c0b8bcfe99ebba2e8b762cf2ab72f32751
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |34 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |3 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |26 sec  (00:00:26) |
|test (hthor) |836 |836 |0 |201 sec  (00:03:21) |
|test (thor) |759 |759 |0 |617 sec  (00:10:17) |
|test (roxie) |912 |912 |0 |205 sec  (00:03:25) |


HPCC Stop: OK

HPCCSmoketest commented:
Process of PR-12606, label: HPCC-22124-fix-master starts now.
The reason of this test is: New commit.
Commit ID: 233c37c0b8bcfe99ebba2e8b762cf2ab72f32751
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 233c37c0b8bcfe99ebba2e8b762cf2ab72f32751
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |3 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |27 sec  (00:00:27) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |24 sec  (00:00:24) |
|test (hthor) |836 |836 |0 |180 sec  (00:03:00) |
|test (thor) |759 |759 |0 |622 sec  (00:10:22) |
|test (roxie) |912 |912 |0 |223 sec  (00:03:43) |


HPCC Stop: OK

HPCCSmoketest commented:
Process of PR-12606, label: HPCC-22124-fix-master starts now.
The reason of this test is: New commit.
Commit ID: 233c37c0b8bcfe99ebba2e8b762cf2ab72f32751
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 233c37c0b8bcfe99ebba2e8b762cf2ab72f32751
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |28 sec  (00:00:28) |
|setup (thor) |11 |11 |0 |47 sec  (00:00:47) |
|setup (roxie) |11 |11 |0 |18 sec  (00:00:18) |
|test (hthor) |836 |836 |0 |158 sec  (00:02:38) |
|test (thor) |759 |759 |0 |627 sec  (00:10:27) |
|test (roxie) |912 |912 |0 |210 sec  (00:03:30) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   29 sec (00:00:29) |  213 sec (00:03:33) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1327 sec (00:22:07) |   19 sec (00:00:19) | 1608 sec (00:26:48) |

HPCCSmoketest commented:
Process of PR-12606, label: HPCC-22124-fix-master starts now.
The reason of this test is: New commit.
Commit ID: 40afe09e1ea5754093dfa2cd86a3954b988fb143
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 40afe09e1ea5754093dfa2cd86a3954b988fb143
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |34 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |27 sec  (00:00:27) |
|setup (thor) |11 |11 |0 |46 sec  (00:00:46) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |836 |836 |0 |194 sec  (00:03:14) |
|test (thor) |759 |759 |0 |629 sec  (00:10:29) |
|test (roxie) |912 |912 |0 |206 sec  (00:03:26) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   29 sec (00:00:29) |  214 sec (00:03:34) |    0 sec (00:00:00) |    3 sec (00:00:03) |   19 sec (00:00:19) | 1342 sec (00:22:22) |   19 sec (00:00:19) | 1626 sec (00:27:06) |

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1199,57 +1198,65 @@ class CDFUengine: public CInterface, implements IDFUengine
+
+                    iskey = isFileKey(srcFile);
+                    if ((cmd==DFUcmd_copy) && (srcFile->querySuperFile() != nullptr) && iskey)
+                        throwError1(DFTERR_InvalidSuperindexCopy, srcFile->queryLogicalName());
```
trivial: as tmp is same as srcFile->queryLogicalName() would be good to be consistent here and on line 1227.
And a littler further down it is set into srcName.
You could set srcName to tmp ahead of the lookup() on line 1223, and then use srcFile in all cases.

AttilaVamos replied:
Changed

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1199,57 +1198,65 @@ class CDFUengine: public CInterface, implements IDFUengine
+                    oldRoxiePrefix.set(srcFile->queryAttributes().queryProp("@roxiePrefix"));
+                    kind.set(srcFile->queryAttributes().queryProp("@kind"));
+
+                    if (destination->getWrap()||(iskey&&(cmd==DFUcmd_copy)))
```
probably worth reinstating comment that was here:
```
    // keys default wrap for copy
```

AttilaVamos replied:
Done

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1199,57 +1198,65 @@ class CDFUengine: public CInterface, implements IDFUengine
-                            destination->setNumPartsOverride(srcFile->numParts());
-                        if (options->getSubfileCopy())
-                            opttree->setPropBool("@compress",srcFile->isCompressed());
+
```
formatting: spurious extra line.

AttilaVamos replied:
REmoved

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1330,7 +1337,7 @@ class CDFUengine: public CInterface, implements IDFUengine
-                                    IFileDescriptor * srcDesc = (auxfdesc.get() ? auxfdesc.get() : foreignfdesc.get());
+                                    IFileDescriptor * srcDesc = (auxfdesc.get() ? auxfdesc.get() : srcFile->getFileDescriptor());
                                     dstCompressed = srcDesc && srcDesc->isCompressed();
                                 }
```
if srcFile were non null it cannot be here..
So I think should change these lines to:
```
   else if (auxfdesc)
       dstCompressed = auxfdesc->isCompressed();
```

AttilaVamos replied:
Changed

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1405,23 +1412,25 @@ class CDFUengine: public CInterface, implements IDFUengine
+                        if (diffNameSrc.get()||diffNameDst.get())
+                        {
                             Owned<IFileDescriptor> newf;
+                            newf.setown(srcFile->getFileDescriptor());
```
trivial: clearer as 1 line, i.e.:
```
    Owned<IFileDescriptor> newf = srcFile->getFileDescriptor();
```

AttilaVamos replied:
Changed

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1447,7 +1456,7 @@ class CDFUengine: public CInterface, implements IDFUengine
                             feedback.repmode=cProgressReporter::REPbefore;
                         if (foreigncopy)
-                            checkPhysicalFilePermissions(foreignfdesc,userdesc,false);
+                            checkPhysicalFilePermissions(srcFile->getFileDescriptor(),userdesc,false);
```
this will leak the IFileDescriptor.

AttilaVamos replied:
Changed

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1486,7 +1495,7 @@ class CDFUengine: public CInterface, implements IDFUengine
                         else if (foreigncopy||auxfdesc)
                         {
-                            IFileDescriptor * srcDesc = (auxfdesc.get() ? auxfdesc.get() : foreignfdesc.get());
+                            IFileDescriptor * srcDesc = (auxfdesc.get() ? auxfdesc.get() : srcFile->getFileDescriptor());
```
leaks the IFileDescriptor if getFileDescriptor called here.

AttilaVamos replied:
Changed

jakesmith commented:
@AttilaVamos - a few minor issues.

HPCCSmoketest commented:
Process of PR-12606, label: HPCC-22124-fix-master starts now.
The reason of this test is: New commit.
Commit ID: 35dbde926d6da7b39dc741f2e4c1c9df738f7c8c
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 35dbde926d6da7b39dc741f2e4c1c9df738f7c8c
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |46 sec  (00:00:46) |
|setup (roxie) |11 |11 |0 |27 sec  (00:00:27) |
|test (hthor) |836 |836 |0 |174 sec  (00:02:54) |
|test (thor) |759 |759 |0 |616 sec  (00:10:16) |
|test (roxie) |912 |912 |0 |219 sec  (00:03:39) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   35 sec (00:00:35) |  208 sec (00:03:28) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1325 sec (00:22:05) |   19 sec (00:00:19) | 1608 sec (00:26:48) |

AttilaVamos commented:
@jakesmith updated, please recheck, especially the srcFdesc and its usage.

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1226,13 +1227,16 @@ class CDFUengine: public CInterface, implements IDFUengine
                         throw MakeStringException(-1,"Source file %s could not be found",tmp.str());
 
+                    srcName.set(tmp.str());
+                    srcFdesc.set(srcFile->getFileDescriptor());
```
that will still leak.
getFilleDescriptor() returns a linked IFileDescriptor.
Owned::set also links.

you should use setown

AttilaVamos replied:
Changed

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1226,13 +1227,16 @@ class CDFUengine: public CInterface, implements IDFUengine
                     if (!srcFile)
                         throw MakeStringException(-1,"Source file %s could not be found",tmp.str());
 
+                    srcName.set(tmp.str());
```
picky: doesn't really need .str() here, the const char * operator, will automatically be invoked.

AttilaVamos replied:
Changed

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1418,8 +1418,7 @@ class CDFUengine: public CInterface, implements IDFUengine
                         {
-                            Owned<IFileDescriptor> newf;
-                            newf.setown(srcFile->getFileDescriptor());
+                            Owned<IFileDescriptor> newf = srcFdesc.get();
```
this is -ve leak (i.e. this will incorrectly Release and ultimately could cause a crash as free too early).

Why have 'newf' at all now? just use srcFDesc.

AttilaVamos replied:
Changed

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1246,20 +1250,18 @@ class CDFUengine: public CInterface, implements IDFUengine
                     }
                     if (destination->getMultiCopy()&&!destination->getWrap())
                     {
                         Owned<IFileDescriptor> tmpfd = srcFile->getFileDescriptor();
```
since you have srcFDesc in scope now, this can be replaced too.

AttilaVamos replied:
Changed

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1614,7 +1613,7 @@ class CDFUengine: public CInterface, implements IDFUengine
                     fsys.replicate(fdesc.get(), mode, recovery, recoveryconn, filter, opttree, &feedback, &abortnotify, dfuwuid);
                     runningconn.clear();
                     if (!abortnotify.abortRequested()) {
-                        Audit("REPLICATE",userdesc,srcFile?srcFile->queryLogicalName():NULL,NULL);
```
line 1611: Owned<IFileDescriptor> fdesc = srcFile->getFileDescriptor();

if srcFile is valid - so is srcFDec, so you can replace this.

jakesmith replied:
line 1580:                        srcFile.clear();

Now you have introduced a srcFDesc at same scope, you should clear it here as well.

jakesmith commented:
@AttilaVamos - please see comments.

HPCCSmoketest commented:
Process of PR-12606, label: HPCC-22124-fix-master starts now.
The reason of this test is: New commit.
Commit ID: fdb2629b1223598a6591900a3f655f4e62273b06
Estimated completion time is ~0.5 hour(s)

AttilaVamos commented:
@jakesmith updated, please re-check.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: fdb2629b1223598a6591900a3f655f4e62273b06
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |34 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |3 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |30 sec  (00:00:30) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |836 |836 |0 |188 sec  (00:03:08) |
|test (thor) |759 |759 |0 |619 sec  (00:10:19) |
|test (roxie) |912 |912 |0 |227 sec  (00:03:47) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   36 sec (00:00:36) |  215 sec (00:03:35) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1348 sec (00:22:28) |   20 sec (00:00:20) | 1640 sec (00:27:20) |

jakesmith commented:
@AttilaVamos - looks good, please squash.

AttilaVamos committed:
HPCC-22124 File services permits and defaults to copying a super index into a invalid single index with multiple TLK's

Add Superfile check for a simple logical file copy.

Add supercopy.ecl to Regression Suite

Remove 'foreignfdesc'

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Process of PR-12606, label: HPCC-22124-fix-master starts now.
The reason of this test is: New commit.
Commit ID: 96666d18e1d874b9bd281a67588f1b670a1513a3
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 96666d18e1d874b9bd281a67588f1b670a1513a3
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |27 sec  (00:00:27) |
|setup (thor) |11 |11 |0 |46 sec  (00:00:46) |
|setup (roxie) |11 |11 |0 |16 sec  (00:00:16) |
|test (hthor) |836 |836 |0 |163 sec  (00:02:43) |
|test (thor) |759 |759 |0 |604 sec  (00:10:04) |
|test (roxie) |912 |912 |0 |216 sec  (00:03:36) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   33 sec (00:00:33) |  213 sec (00:03:33) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1313 sec (00:21:53) |   20 sec (00:00:20) | 1599 sec (00:26:39) |
